Cache verification of unchanged methods

JonasAlaif opened the pull request:
This caching will be implemented at the `VerificationRequest` level

JonasAlaif committed:
Implement `Hash` for `VerificationRequest`

JonasAlaif committed:
Avoid using UUID in forall variable names

JonasAlaif committed:
Make Clippy and fmt happy

JonasAlaif committed:
Make sure `Expr` implements PartialEq

JonasAlaif committed:
Make large parts of poly-vir `Ord` to allow sorting

JonasAlaif committed:
Clippy and fmt

JonasAlaif committed:
Fix fmt issue

I should really start doing these checks locally...

JonasAlaif committed:
Improve stability of closure and fndef encoding

vakaras commented on prusti-viper/src/encoder/mir/pure/specifications/encoder.rs:
```diff
@@ -255,7 +255,7 @@ impl<'p, 'v: 'p, 'tcx: 'v> SpecificationEncoder<'p, 'v, 'tcx> {
                                         self.encode_quantifier_arg(
                                             *arg,
                                             arg_ty,
-                                            &format!("{}_{}", vars.spec_id, vars.pre_id),
```
Have you checked whether this does not lead to strange name conflicts in encoded programs that are insanely hard to understand on the source level?

vakaras replied:
It would be great to have a bunch of tests that check this.

JonasAlaif replied:
I think that might be possible; I'll take a look at creating some test cases for this. However, in general I think this is very hard to avoid with deterministic names - one could just look at the generated Viper code and name another local variable in a way which clashes. The two solutions to avoid this that I see: (A) use hash of the HIR/MIR of the function as a sort of 'spec_id' which could be used in these cases (then if another variable is defined after the fact, this hash will change) or (B) look at all defined variables in scope and pick a name that doesn't clash.

vakaras replied:
> (B) look at all defined variables in scope and pick a name that doesn't clash.

We could do this when lowering to `vir_legacy` – at that point, we have all variable names present.

fpoli replied:
> one could just look at the generated Viper code and name another local variable in a way which clashes

Would that still be possible if we add a prefix `quantified_` (or just `q_`) to all quantified variables?

Even with that, as Vytautas asked I wonder what happens to the encoding of `forall x: u32 :: forall x: i32 :: ...`. Will the Viper names clash?

JonasAlaif replied:
Aha I forgot that the user doesn't have control over MIR names. In that case we probably don't need to resort to the drastic measures I mention above? I'll think about where this might not hold

Aurel300 replied:
FYI in the spec refactor I base this ID on the `DefId` of the closure it comes from.

JonasAlaif replied:
@Aurel300 do you mean the `spec_id`? The `pre_id` is just the level of nesting and shouldn't change in the refactor, right?

Aurel300 replied:
In my WIP branch, quantifier variables are named with `format!("_{}_quant_{}", arg_idx, body_def_id.index.index())`, where `arg_idx` is the position in the quantifier's arguments and `body_def_id` is the `DefId` of the closure containing the quantifier body. It is a local `DefId`, so `index` is the only number we use here.

JonasAlaif replied:
@Aurel300 Hmm, but using the `DefId` is problematic in this case since it won't be stable if other stuff in the file/crate is added/deleted. I'm not exactly sure what position in quantifiers args (`arg_idx`) means (do you mean that for a `forall(|a: i32, b: i32| forall(|c: i32| ...))`, then `a` is 1, `b` is 2, `c` is 1?), but could we not use only this, with also a `quantifier_nesting_count` in the name as well?

Aurel300 replied:
Yes, it wasn't chosen with caching in mind, and it is still better than the original approach of random UUIDs. A nesting count would work, but I'm not sure it would be easy to introduce? When encoding specifications into Viper expressions we would generally work inside out, so it's hard for the inner `forall` to know its nesting depth.

JonasAlaif replied:
The current `vars.pre_id` seems to do this now though (at least from looking at the generated Viper), I'm not sure how it's implemented but maybe if we just copied the ideas from that?

Aurel300 replied:
I thought `vars.pre_id` corresponded to another specification UUID? Maybe it's being interned somewhere?

JonasAlaif committed:
Make quantifiers more stable

JonasAlaif committed:
Add two Prusti flags for debugging

`print_hash` skips verification and just print the hash of the verification request. `disable_cache` will prevent caching to enable the debugging of performance. Both should be documented in the manual

JonasAlaif committed:
Cache requests and save results to disk

JonasAlaif committed:
Add a test which checks that the hash remains stable

JonasAlaif committed:
Revert change which was accidentally included

This was to test if the error reporting still works with caching - it works

JonasAlaif committed:
Clippy fix

fpoli commented on prusti-server/src/process_verification.rs:
```diff
@@ -8,12 +8,45 @@ use crate::{VerificationRequest, ViperBackendConfig};
+        let hash = request.get_hash();
+        // Try to load from cache and return `result`
+        if let Some(result) = cache.get(hash) {
+            return result;
```
```suggestion
            info!("Using verification result from the cache");
            return result;
```

Just to make debugging easier in the IDE, when users include their log when reporting issues.

`info!("Verification request hash: {}", hash);` two lines above might be useful too.

fpoli commented on prusti-server/src/process_verification.rs:
```diff
@@ -8,12 +8,45 @@ use crate::{VerificationRequest, ViperBackendConfig};
+    // Print hash of the `VerificationRequest`
+    if config::print_hash() {
+        println!(
+            "Recieved verification request for: {}",
```
```suggestion
            "Received verification request for: {}",
```

fpoli commented on prusti-server/src/server.rs:
```diff
@@ -54,35 +54,36 @@ where
     stopwatch.finish();
 
-    let build_verification_request_handler = |viper_arc: Arc<Viper>| {
+    let cache_loc = config::cache_dir() + "data.json";
```
`let cache_loc = PathBuf::from(config::cache_dir()).join("data.json");` is more robust, e.g. if `cache_dir` doesn't end with a `/`.

fpoli commented on prusti-server/src/server.rs:
```diff
@@ -54,35 +54,36 @@ where
+    let build_verification_request_handler = |viper_arc: Arc<Viper>, cache| {
         move |request: VerificationRequest| {
             // Here it would be easy to cache the result of verification requests, as soon as
             // `VerificationRequest` implements `Hash`.
```
```suggestion
```

fpoli commented on prusti-viper/src/verifier.rs:
```diff
@@ -422,8 +422,10 @@ fn verify_programs(env: &Environment, programs: Vec<vir::Program>)
         stopwatch.start_next("attach current thread to the JVM");
         let viper_thread = viper.attach_current_thread();
         stopwatch.finish();
+        let cache_loc = config::cache_dir() + "data.json";
```
`PathBuf::from(...).join(...)` is more robust.

Also, what about renaming `cache_dir` to `cache_path`, making the default `"cache/data.json"`? That way, the user (or the IDE) can use e.g. `/tmp/prusti_cache.json`.

fpoli commented on prusti-viper/tests/test_caching.rs:
```diff
@@ -0,0 +1,105 @@
+        assert!(out.status.success(), "Failed to compile: {:?}\n{}", program, String::from_utf8(out.stderr).unwrap());
+        let stdout = String::from_utf8(out.stdout).unwrap();
+        let mut hash_lines = stdout.lines()
+            .skip_while(|line| !line.starts_with("Recieved verification request for: "));
```
```suggestion
            .skip_while(|line| !line.starts_with("Received verification request for: "));
```

fpoli commented on prusti-viper/tests/test_caching.rs:
```diff
@@ -0,0 +1,105 @@
+        let mut hash_lines = stdout.lines()
+            .skip_while(|line| !line.starts_with("Recieved verification request for: "));
+        while let Some(l1) = hash_lines.next() {
+            let full_name = l1.strip_prefix("Recieved verification request for: ").unwrap();
```
```suggestion
            let full_name = l1.strip_prefix("Received verification request for: ").unwrap();
```

fpoli commented on viper/src/verification_result.rs:
```diff
@@ -70,3 +78,82 @@ pub struct JavaExceptionWithOrigin {
+    fn insert(self, request: u64, result: VerificationResult) -> Option<VerificationResult>;
+}
+#[derive(Debug, Clone)]
+pub struct CacheData {
```
What about `PersistentCache` instead of `CacheData`?

Also, I'd move all the cache-related code to a new `src/cache.rs` file.

fpoli commented on viper/src/verification_result.rs:
```diff
@@ -70,3 +78,82 @@ pub struct JavaExceptionWithOrigin {
+#[derive(Debug, Clone)]
+pub struct CacheData {
+    updated: bool,
+    load_loc: String,
```
`PathBuf` instead of `String`?

fpoli commented on prusti-common/src/config.rs:
```diff
@@ -101,6 +102,8 @@ lazy_static! {
         settings.set_default("hide_uuids", false).unwrap();
         settings.set_default("counterexample", false).unwrap();
+        settings.set_default("print_hash", false).unwrap();
+        settings.set_default("disable_cache", false).unwrap();
```
`enable_cache` might be more immediate (one less negation = one less neuron required)

Pointerbender commented on prusti-viper/tests/hash/layout_1.rs:
```diff
@@ -0,0 +1,157 @@
+use prusti_contracts::*;
+
+// Borrows are not stable since rustc::Loan indicies aren't
+// i.e. the `N` used in `DeadBorrowToken$(N)` for wands
+// will differ between compilation
+// This includes fns which return borrows, or e.g. slices
+// where the fold-unfold algorithm doesn't handle the exhale well
```
Are the loan indices unstable because of how Polonius returns the indices, or are these renumbered in a non-deterministic fashion somewhere in Prusti? (e.g. due to the use of a randomly seeded HashSet/HashMap somewhere).

Pointerbender replied:
Indeed this turned out to be caused by the use of randomly seeded `HashSet`/`HashMap` in various places. PR #817 to the `master` branch should solve this!

JonasAlaif committed:
Merge branch 'pr/817' into stable-viper-for-caching

JonasAlaif committed:
Implement review changes

JonasAlaif committed:
Merge branch 'master' into stable-viper-for-caching

JonasAlaif committed:
Fix merge issues

JonasAlaif committed:
Clippy fix

JonasAlaif committed:
Disable name interning until it is improved

JonasAlaif committed:
Update error annotation

JonasAlaif committed:
Update error annotations on verify_partial tests

JonasAlaif committed:
Improve method naming

JonasAlaif committed:
Merge branch 'master' into stable-viper-for-caching

JonasAlaif committed:
Reenable name interner, but ignore method name in hash

JonasAlaif committed:
Merge branch 'stable-viper-for-caching' of https://github.com/JonasAlaif/prusti-dev into stable-viper-for-caching

JonasAlaif committed:
Fix expected error and encode * in identifier

JonasAlaif committed:
Merge branch 'master' into stable-viper-for-caching

JonasAlaif committed:
Fix caching and tests on windows

JonasAlaif committed:
Disable caching by default until the IDE can reset it

JonasAlaif committed:
Add debug info when loading and saving cache

JonasAlaif committed:
Enable memory only caching by default, and only save to disk if `cache_path` is set

JonasAlaif committed:
Switch to saving cache as a binary file and improve error reporting

The binary file should take up much less space than a json, and can easily use compression in the future if required

fpoli commented on viper/src/cache.rs:
```diff
@@ -0,0 +1,126 @@
+                        data_res = Some(data);
+                    }
+                    Err(e) =>
+                        error!("Failed to read cache from \"{}\": {e}", cache_loc.display()),
```
Small nit: if the file doesn't exist we can omit this error.

fpoli replied:
Oh, but then `open` fails. Ignore my comment.

JonasAlaif replied:
Yeah, I had a more detailed structure where the first two `if`s would produce `info!` output in the `else` cases but then removed that in order to simplify things and instead just use the generic `info!("Cache file doesn't exist or is invalid. Using fresh cache.")` thing at the end.
But yes, this error here is if we try to load the cache from some file that has some random data which we fail to parse.

JonasAlaif committed:
Clippy and fmt fixes

JonasAlaif committed:
Add save cache endpoint to prusti-server
